sky/provision/gcp/config.py

```diff
@@ -729,12 +748,58 @@ def get_gpu_direct_usable_vpcs_and_subnets(
                                     compute,
                                     network=vpc_name)
         # Apply firewall rules
-        _create_rules(project_id, compute, constants.FIREWALL_RULES_TEMPLATE,
-                      vpc_name)
+        # No need to recreate the rules if exist,
+        # as they are totally managed by SkyPilot,
+        # in this case, we can skip the rules creation during failover
+        _create_rules(project_id,
+                      compute,
+                      constants.FIREWALL_RULES_TEMPLATE,
+                      vpc_name,
+                      recreate=False)
         vpc_subnet_pairs.append((vpc_name, subnets[0]))
     return vpc_subnet_pairs
 
 
+def delete_gpu_direct_vpcs_and_subnets(
+    cluster_name: str,
+    project_id: str,
+    region: str,
+    failover: bool = False,
+):
+    """Delete GPU Direct subnets, firewalls, and VPCs."""
+    compute = _create_compute()
+    cluster_prefix = cluster_name
+
+    # TODO(hailong): Determine the num_vpcs per different GPU Direct types
+    num_vpcs = constants.SKYPILOT_GPU_DIRECT_VPC_NUM
+
+    for i in range(num_vpcs):
+        if i == 0:
+            vpc_name = f'{cluster_prefix}-mgmt-net'
+        else:
+            vpc_name = f'{cluster_prefix}-data-net-{i}'
+        # Check if VPC exists
+        vpc_list = _list_vpcnets(project_id, compute, filter=f'name={vpc_name}')
+        if not vpc_list:
+            continue
+        for vpc in vpc_list:
+            subnets = _list_subnets(project_id,
+                                    region,
+                                    compute,
+                                    network=vpc['name'])
+            for subnet in subnets:
+                logger.info(f'Deleting subnet {subnet["name"]}')
+                _delete_subnet(project_id, region, compute, subnet['name'])
+
+            if not failover:
+                # For failover, we don't delete the rules and VPCs,
+                # which are global resources and can be reused.
```

Comment:
We could be failing over from GCP to another cloud. In that case it seems like we should clean up the global resources. Do we handle this case?

Reply:
We could launch with `network_tier: best`, which could apply to multiple clouds, right? At the moment, only GCP has support for `network_tier`, but other clouds could be added later. Also, it seems the user could failover between completely different resources, like:
```
resources:
  ordered:
    - infra: gcp
      network_tier: best
    - infra: aws
```
Just want to make sure this does lead to a leak. Maybe I am misunderstanding though.

Reply:
Yes, the VPCs will be leaked in this case. Is it OK that I create an issue for this and address this in a separate PR to make the cleanup work for most cases first?